<a href="https://colab.research.google.com/github/jardinthee/codetree-TILs/blob/main/deep_13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import pandas as pd

!git clone https://github.com/taehojo/data.git

df = pd.read_csv('./data/sonar3.csv', header=None)


df.head()

fatal: destination path 'data' already exists and is not an empty directory.


0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   0  
1  0.0052  0.0044   0  
2  0.0095  0.0078   0  
3  0.0040  0.0117   0  
4  0.0107  0.0094   0  

[5 rows x 61 columns]

In [5]:
df[60].value_counts()

1    111
0     97
Name: 60, dtype: int64

위 value_counts() 값을 보면 광석이 111개, 일반 암석이 97개가 있음을 알 수 있다

In [9]:
X = df.iloc[:,0:60]
y = df.iloc[:,60]

In [7]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [10]:
model = Sequential()
model.add(Dense(24, input_dim=60,activation='relu'))
model.add(Dense(10, activation = 'relu'))
model.add(Dense(1, activation = 'sigmoid'))

model.compile(loss='binary_crossentropy' , metrics = ['accuracy'])

history = model.fit(X,y,epochs=200,batch_size=10)


Epoch 1/200
21/21 [==============================] - 1s 2ms/step - loss: 0.6901 - accuracy: 0.5337
Epoch 2/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6802 - accuracy: 0.5240
Epoch 3/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6684 - accuracy: 0.5913
Epoch 4/200
21/21 [==============================] - 0s 4ms/step - loss: 0.6590 - accuracy: 0.6298
Epoch 5/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6525 - accuracy: 0.6202
Epoch 6/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6442 - accuracy: 0.6538
Epoch 7/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6340 - accuracy: 0.6587
Epoch 8/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6211 - accuracy: 0.6779
Epoch 9/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6113 - accuracy: 0.6683
Epoch 10/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6014 - accuracy: 0.6587
Epoch 11/

**과적합 방지**

In [12]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split

import pandas as pd

X = df.iloc[:,0:60]
y = df.iloc[:,60]

#학습셋과 데이터 셋을 구분합니다
X_train , X_test , y_train, y_test = train_test_split(X, y, test_size=0.3, shuffle=True)

model = Sequential()
model.add(Dense(24,input_dim = 60, activation='relu'))
model.add(Dense(10,activation='relu'))
model.add(Dense(1,activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam',metrics=['accuracy'])

history = model.fit(X_train, y_train, epochs=200, batch_size=10)

score = model.evaluate(X_test, y_test)

print('Test accuracy:' , score[1])

Epoch 1/200
15/15 [==============================] - 1s 2ms/step - loss: 0.6932 - accuracy: 0.5172
Epoch 2/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6705 - accuracy: 0.5931
Epoch 3/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6587 - accuracy: 0.6552
Epoch 4/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6472 - accuracy: 0.7034
Epoch 5/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6379 - accuracy: 0.7034
Epoch 6/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6269 - accuracy: 0.7310
Epoch 7/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6171 - accuracy: 0.7103
Epoch 8/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6057 - accuracy: 0.7793
Epoch 9/200
15/15 [==============================] - 0s 2ms/step - loss: 0.5984 - accuracy: 0.7448
Epoch 10/200
15/15 [==============================] - 0s 2ms/step - loss: 0.5865 - accuracy: 0.7586
Epoch 11/